Create simple version of ratelimit package

justinsb opened the pull request:
Allows for better testing.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35583)

<!-- Reviewable:end -->

justinsb commented:
@k8s-bot cvm gce e2e test this

justinsb commented:
@k8s-bot gci gce e2e test this

justinsb commented:
@k8s-bot gce etcd3 e2e test this

jbeda commented:
Awesome!  Happy to review this.

Quick request -- can you break out the godep/vendor stuff in to a separate commit?  It helps to keep the "real code" easy to review.

jbeda reviewed (commented):
Looks pretty good.  Mostly nits around comments.

justinsb committed:
Remove juju ratelimit from godeps

justinsb committed:
Create simple version of ratelimit package

Allows for more testing.

justinsb commented on pkg/util/ratelimit/bucket.go:
```diff
@@ -0,0 +1,154 @@
+// Bucket models a token bucket
+type Bucket struct {
+	unitsPerNano float64
+	nanosPerUnit float64
```
Will do - I also hand-waving satisifed myself that there's no loss of precision for qps from about 1E6 to 1E-6.

I'll actually add some unit tests just to prove it :-)

justinsb commented:
The "just one more unit test" found a bug actually ... pathological rates like (1E9 + 1) were a bit tricky, so I added a "fractional capacity" field which holds the non-integer capacity we added, which makes the time math much simpler.

justinsb commented:
@k8s-bot node e2e test this

luxas commented:
Oh, how sweet isn't this?
Removing 1k LoC is always fun :)
:+1:

Will we use this ratelimit package when we're gonna limit the unauthorized access to the `kube-public` namespace?

jbeda commented:
I'm really not sure why this hasn't been merged.  Let me toggle LGTM to see if that does it.

luxas commented:
It's just because the Submit-Queue is sooo long unfortunately.

If this is urgent, add priority/P2 and it will merge quite soon

smarterclayton commented:
Changes like this should have included review from sig-API-machinery

mikedanese commented:
FYI, there is a ratelimiter in [golang.org/x/time](https://godoc.org/golang.org/x/time/rate).

liggitt commented:
not opposed to switching if there's a shortcoming of the one in use, but stuff under golang.org/x/... still has to be vendored in, and that once doesn't appear significantly simpler or more capable for our needs

mikedanese commented:
From my understanding, the main problems we've (google) had around the juju ratelimiter is the license. We had to hack around it to be able to use the client internally because lawyers. I don't understand the "allows for better testing" but would like to understand the juju ratelimter short-comings better.

liggitt commented:
looks like golang.org/x/ratelimit is close, it's just missing an Available() method we need to calculate saturation for metrics reporting. Opened https://github.com/golang/go/issues/18233
